README.md
# ech0raix_decryptor
  
If you are here it means that your NAS has been infected by ech0raix (aka qnap encrypt) and that you 
are either :  
  
- searching for a way to decrypt your files without paying and it might be possible if the ransomware used to encrypt your data is from 2019.  
  Generally 2019 malware creates a text file named **README_FOR_DECRYPT.txt** and some developpers have released a brute force decryptor.
  However if you have been infected by a more recent versions (in this case the text file is **README_FOR_DECRYPT.txtt**) there is no other option than to pay.
  Hope on the futur we will find a way of decoding it.  
    
- You have paid and you have received the decryptor binaries but you don't trust them and you want to execute virus free binary.  
  So welcome on this page and please follow the instruction:
  
  **Step 1)** You first need to extract the key embedded inside the "official" (it's hard to use this word) decryptor_binary.  
              To do so you can download Free version of IDA Disassembler and search inside main_main to find the key as shown below:                 
              ![Alt text](https://github.com/vricosti/ech0raix_decryptor/blob/main/doc/ida_extract_key.png?sanitize=true)

  **Step 2)** Try first to run the decryptor on a folder without deleting the encrypted files.  




blood_dolly_decryptor/changelog.txt
Changelog for ECh0raix Decoder
--------------------------------

**********************
* ECh0raix Decoder *
**********************
=========
= 1.0.6 =
=========
- Improved handling of paths longer than 260 characters

=========
= 1.0.5 =
=========
- Improved repair algorithm for encrypted files bigger than 20MB (unecrypted size bigger than 10MB)

=========
= 1.0.4 =
=========
- Added support to use file as a source of decryption keys (Check 4.3. in readme)

=========
= 1.0.3 =
=========
- Speed up bruteforce on CPUs with AES hardware support and SSSE3
- Added exhaustive search (just for fun)

=========
= 1.0.2 =
=========
- Fixed a bug that prevented to load correct found key from brute forcer

=========
= 1.0.1 =
=========
- Fixed a bug that prevented to decrypt files bigger than 100MB
- Added support for obtaining the key from the ransom note if RSA private key is known

=========
= 1.0.0 =
=========
- First public release

findfile.py
import sys
import os
import platform
import datetime


def creation_date(path_to_file):
    """
    Try to get the date that a file was created, falling back to when it was
    last modified if that isn't possible.
    See http://stackoverflow.com/a/39501288/1709587 for explanation.
    """
    if platform.system() == 'Windows':
        return os.path.getctime(path_to_file)
    else:
        try:
            stat = os.stat(path_to_file)
            try:
                return stat.st_birthtime
            except AttributeError:
                # We're probably on Linux. No easy way to get creation dates here,
                # so we'll settle for when its content was last modified.
                return stat.st_mtime
        except FileNotFoundError:
            return None


def main(start_path):
    if start_path == '.' or start_path == './':
        cur_path = os.path.dirname(os.path.realpath(__file__))
        start_path = cur_path

    for dirpath, subdirs, files in os.walk(start_path, followlinks=True):
         for file in files:
             filepath = os.path.join(dirpath, file)
             if os.path.exists(filepath):
                date = creation_date(filepath)
                if date:
                    dt = datetime.datetime.fromtimestamp(date)
                    if not file.endswith('.encrypt'):
                        if dt.day == 14 and dt.month == 6 and dt.year == 2022:
                            print("Found file with corresponding date {}".format(filepath))


if __name__ == "__main__":
    if len(sys.argv) != 2:
        sys.exit("Missing path argument")

    main(sys.argv[1])

blood_dolly_decryptor/Ech0raixDecoder_Decrypt.asm
;4STDs9cmUlkiujXuLkdTouoqOIfER4TE
;D:\Dev\RansomWare\Test
; They had a problem in their decryptor (maybe they still have) what corrupted files bigger than 10MB.
; This ransomware has plenty of issues and one of them is how they handle file encryption. 
; The first odd thing is that they write encrypted bytes 2 times, so encrypted file is 2 times bigger + IV for each block + "ech0raix" marker of course. 
; Another odd thing is that during encryption they reset IV.
; So first what you have to check is if at offset 10*1024*1024 + 16 is the same IV as at the start of the encrypted file. 
; If yes, then you have to reset IV each 10 MB period. If the period is not there, than it is older version and period is 100MB instead of 10MB. 
; Maybe they changed it or fixed their decryptor, but as I said I did not recheck their ransomware nor decryptor for long time. 
; After you handle the period corrently you just have to check if there is "ech0raix" marker in decrypted buffer. 
; You can calculate the original filesize of course, but it is more bullet proof to just check it - more resistent for future changes. 
; That is the reason why I am doing it like that.
;------------------------------------------------------------------------------------------------------------------------------------------------
; FILEBUFFERSIZE equ 10485760
;------------------------------------------------------------------------------------------------------------------------------------------------
align 16
_DecryptFile proc uses esi edi ebx ecx edx lpFilename:DWORD, lpFindData:DWORD, deleteorig:DWORD, version:DWORD

        local hfile_in:DWORD
        local hfile_out:DWORD
        local _exitvalue:DWORD
        local bytesread:DWORD
        local size_low:DWORD
        local size_high:DWORD
        local warning_level:DWORD
        local _nKey:DWORD
        local lpkey:DWORD
        local filecreated:DWORD
        local lpAES:AES
        local _IV[16]:BYTE
        local marker:DWORD
        local echoraixpos:DWORD
        local block_100:DWORD
        local keyfiletested:DWORD
        local period:DWORD

        xor eax,eax
        mov hfile_in,eax
        mov hfile_out,eax
        mov _exitvalue,eax
        mov bytesread,eax
        mov warning_level,eax
        mov _nKey,-1
        mov lpkey,eax
        mov filecreated,eax
        mov marker,eax
        mov block_100,eax
        mov keyfiletested,eax
        mov period,100*1024*1024
        
       ;check version
        mov eax,version
        test eax,eax
        jz _Decrypt_invalidversion
        cmp eax,VERSIONS
        ja _Decrypt_invalidversion
        
       ;open encrypted file
        invoke CreateFile,lpFilename,GENERIC_READ,FILE_SHARE_READ,0,OPEN_EXISTING,0,0
        cmp eax,-1
        jz _Decrypt_open_err
        mov hfile_in,eax
        
        invoke strlength,lpFilename
        sub eax,8
        push eax
        push lpFilename
        invoke Pis,heap_cesta2,lpstr("$u")

        mov esi,file_mem_in
        mov edi,file_mem_out
        
       ;copy size
        mov edx,lpFindData
        mov eax,(WIN32_FIND_DATAW ptr [edx]).nFileSizeLow
        mov ecx,(WIN32_FIND_DATAW ptr [edx]).nFileSizeHigh
        and eax,not(15)
        mov size_low,eax
        mov size_high,ecx
        
       ;create new file
        invoke CreateFile,heap_cesta2,GENERIC_READ+GENERIC_WRITE,FILE_SHARE_READ,0,CREATE_ALWAYS,0,0
        cmp eax,-1
        jz _Decrypt_open2_err
        mov hfile_out,eax
        
       ;Get period
        ;Read first 16 bytes
        invoke ReadFile,hfile_in,addr _IV,16,addr bytesread,0
        cmp bytesread,16
        jnz _Decrypt_read_err

        mov bytesread,0
        invoke SetLastError,NO_ERROR
        invoke SetFilePointer,hfile_in,10*1024*1024+16,addr bytesread,FILE_BEGIN
        invoke GetLastError
        test eax,eax
        jnz _Decrypt_gotperiod
        
        invoke ReadFile,hfile_in,esi,16,addr bytesread,0
        cmp bytesread,16
        jnz _Decrypt_gotperiod
        
        invoke datacmp,esi,addr _IV,16
        test eax,eax
        jz _Decrypt_gotperiod
        mov period,10*1024*1024

_Decrypt_gotperiod:
;;DEBUG
__DEBUG lpstr("File: $s, period: $d"),lpFilename,period
;;/DEBUG   

        mov bytesread,0
        invoke SetFilePointer,hfile_in,0,addr bytesread,FILE_BEGIN
        cmp keyfile_mem,0
        jnz _Decrypt_morekeys
        
_Decrypt_findkey:
      ;find correct key from loaded keys
        mov ebx,heap_keys
;        invoke DataCpyN,addr _IV,addr (KEY ptr [ebx]).IV,16
        invoke AES_SetEncryptKey,ebx,addr lpAES
      
        cmp nKeys,1
        jz _Decrypt_readIV

_Decrypt_morekeys:
      ;read marker data
        push ebx
        mov ebx,lpFindData
        mov ecx,period
        mov eax,(WIN32_FIND_DATAW ptr [ebx]).nFileSizeLow
        mov edx,(WIN32_FIND_DATAW ptr [ebx]).nFileSizeHigh
        lea ecx,[ecx*2+32]
        div ecx
        mov ecx,edx
        shr ecx,1
        mov eax,(WIN32_FIND_DATAW ptr [ebx]).nFileSizeLow
        mov edx,(WIN32_FIND_DATAW ptr [ebx]).nFileSizeHigh
        sub eax,ecx
        sbb edx,0
        sub eax,31
        sbb edx,0
        mov ecx,eax
        and ecx,not(15)
        and eax,15
        mov echoraixpos,eax
        mov bytesread,edx
        pop ebx
        invoke SetFilePointer,hfile_in,ecx,addr bytesread,FILE_BEGIN
        
        invoke ReadFile,hfile_in,esi,48,addr bytesread,0
        cmp bytesread,48
        jnz _Decrypt_read_err
        xor ecx,ecx
        mov bytesread,ecx
        invoke SetFilePointer,hfile_in,ecx,addr bytesread,FILE_BEGIN
        
        cmp keyfiletested,0
        jnz _Decrypt_check_setkeys
        
        mov ecx,keyfile_size
        test ecx,ecx
        jz _Decrypt_check_setkeys
        
       ;check keyfile
        mov ebx,keyfile_mem
        mov eax,keyfile_key
        mov edx,echoraixpos
        add edx,edi
        test eax,eax
        jz _Decrypt_test_keyfile_loop
        
       ;test last key from keyfile first
        mov ebx,eax
        invoke AES_SetEncryptKey,ebx,addr lpAES
        invoke AES_CFB_Decrypt,addr [esi+16],edi,32,esi,addr lpAES
        invoke datacmp,edx,offset echoraix_marker,15  
        test eax,eax
        jnz _Decrypt_keyfile_match
        mov ebx,keyfile_mem
        
align 16
_Decrypt_test_keyfile_loop:
        invoke AES_SetEncryptKey,ebx,addr lpAES
        invoke AES_CFB_Decrypt,addr [esi+16],edi,32,esi,addr lpAES
        
        invoke datacmp,edx,offset echoraix_marker,15  
        test eax,eax
        jnz _Decrypt_keyfile_match
        
        inc ebx
        dec ecx
        cmp ecx,32
        jae _Decrypt_test_keyfile_loop
        
       ;if key is not found and there is more keys, test them or load first key from file and go to decrypt
        mov keyfiletested,1
        cmp nKeys,0
        jnz _Decrypt_check_setkeys
        invoke AES_SetEncryptKey,keyfile_mem,addr lpAES
        jmp _Decrypt_readIV
_Decrypt_keyfile_match:
        mov keyfile_key,ebx
      ;/find correct key from loaded file
        jmp _Decrypt_readIV
        
        
       ;check set keys 
align 16
_Decrypt_check_setkeys:
        mov ecx,nKeys
        mov eax,nKey
        mov edx,echoraixpos
        mov _nKey,eax
        add edx,edi
        test eax,eax
        jz _Decrypt_test_key_loop
        
       ;test last key first
        shl eax,5
        lea ecx,[eax+ebx]
        invoke AES_SetEncryptKey,ecx,addr lpAES
        invoke AES_CFB_Decrypt,addr [esi+16],edi,32,esi,addr lpAES
        mov ecx,nKey
        invoke datacmp,edx,offset echoraix_marker,15  
        inc ecx
        test eax,eax
        jnz _Decrypt_key_match
        mov ecx,nKeys
        
       ;test all keys
align 16
_Decrypt_test_key_loop:
        invoke AES_SetEncryptKey,ebx,addr lpAES
        invoke AES_CFB_Decrypt,addr [esi+16],edi,32,esi,addr lpAES
        
        invoke datacmp,edx,offset echoraix_marker,15  
        test eax,eax
        jnz _Decrypt_key_match
        
        add ebx,sizeof(KEY)
        dec ecx
        jnz _Decrypt_test_key_loop
        
       ;if key is not found just load the first key and do not change number of last used key
        invoke AES_SetEncryptKey,heap_keys,addr lpAES
        jmp _Decrypt_readIV
_Decrypt_key_match:
        dec ecx
        mov nKey,ecx
      ;/find correct key from loaded keys

align 16
_Decrypt_readIV:
        invoke ReadFile,hfile_in,addr _IV,16,addr bytesread,0
        cmp bytesread,16
        jnz _Decrypt_read_err
        
        sub size_low,16                         ; size_low = size_low - 16
        sbb size_high,0                         ; size_high = size_high - 0 - C
        mov block_100,16                        ; block_100 = 16
        
align 16        
_Decrypt_loop:
;;DEBUG
__DEBUG lpstr("FileSizeLow: $d, FileSizeHigh: $d"),size_high,size_low
;;/DEBUG  
        xor eax,eax                             ; eax = 0
        mov ebx,FILEBUFFERSIZE                  ; ebx = FILEBUFFERSIZE
        cmp size_high,eax                       ; test = cmp(size_high, 0)
        ja @F                                   ; if test > 0 goto @F
        mov eax,size_low                        ; eax = size_low
        test eax,eax                            ; if size_low == 0
        jbe _Decrypt_loop_end                   ; goto _Decrypt_loop_end
        cmp ebx,eax                             ; test = compare(ebx, size_low)
        cmova ebx,eax                           ; if (ebx > size_low) { ebx = size_low; }
@@:    
       ;align read buffer if needed
        mov eax,block_100                       ; eax = 16
        add eax,ebx                             ; eax = 16 + ebx 
        cmp eax,period                          ; test = cmp(eax, period)
        jbe _Decrypt_read                       ; if eax == period goto _Decrypt_read
       
       ;how many bytes can we read?
        sub eax,period                          ; eax = eax - period
        sub eax,16                              ; eax = eax - 16
        sub ebx,eax                 ;nbytes to period ie ebx = ebx - eax <=> ebx = eax - period - 16
        
_Decrypt_read:    
       ;read data from input file
        invoke ReadFile,hfile_in,esi,ebx,addr bytesread,0
;;DEBUG
__DEBUG lpstr("Going to read: $d, Read: $d"),ebx,bytesread
;;/DEBUG

        cmp bytesread,ebx
        jz @F
        invoke GetLastError
        test eax,eax
        jnz _Decrypt_read_err     
@@:
             
       ;decrypt
        invoke AES_CFB_Decrypt,esi,edi,ebx,addr _IV,addr lpAES
        test eax,eax
        jz _Decrypt_aes_err

       ;search for ECh0raix marker
        invoke datafind,edi,ebx,lpastr("eCh0raix"),8
        test eax,eax
        jz @F
       ;correction of original filesize
        sub eax,edi
        lea ebx,[eax-7]
        mov marker,1
@@:        
       ;write decrypted data 
        invoke WriteFile,hfile_out,edi,ebx,addr bytesread,0
        test eax,eax
        jz _Decrypt_write_err
        cmp bytesread,ebx
        jnz _Decrypt_write_err
        
       ;prepare IV
        invoke DataCpyN,addr _IV,addr [esi+ebx-16],16

        cmp marker,1
        jz _Decrypt_finish
        mov ecx,period
        add block_100,ebx
        add ecx,16
        cmp block_100,ecx
        jnz @F
        mov bytesread,0
        invoke SetFilePointer,hfile_in,ecx,addr bytesread,FILE_CURRENT
        invoke ReadFile,hfile_in,addr _IV,16,addr bytesread,0
        cmp bytesread,16
        jnz _Decrypt_read_err
        mov eax,period
        add eax,16
        sub size_low,eax
        sbb size_high,0
        mov block_100,16
@@:     
        sub size_low,ebx
        ja _Decrypt_loop
        sbb size_high,0
        jmp _Decrypt_loop
        
align 16
_Decrypt_loop_end:
       ;marker was not found during decryption
        inc warning_level
        push heap_cesta2
        invoke Pis,file_mem_out,lpstr("WARNING - Marker not found. Keeping original file. Decrypted file is probably corrupted: $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_finish

_Decrypt_finish:
        mov edx,lpFindData
        invoke SetFileTime,hfile_out,addr (WIN32_FIND_DATAW ptr [edx]).ftCreationTime,addr (WIN32_FIND_DATAW ptr [edx]).ftLastAccessTime,addr (WIN32_FIND_DATAW ptr [edx]).ftLastWriteTime

        invoke GetFileAttributes,lpFilename
        cmp eax,-1
        jz @F
        invoke SetFileAttributes,heap_cesta2,eax
@@:
        mov _exitvalue,1
        cmp deleteorig,1
        jnz @F
        cmp warning_level,0
        jnz @F
        invoke CloseHandle,hfile_in
        invoke CloseHandle,hfile_out
        mov hfile_in,0
        mov hfile_out,0
        invoke DeleteFile,lpFilename
@@:        
        cmp warning_level,0
        jnz @F
        push heap_cesta2
        invoke Pis,file_mem_out,lpstr("DECRYPTED: $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        inc files_decrypted
@@:
        cmp warning_level,0
        jz @F
        inc files_warning
@@:

_Decrypt_end:
        cmp hfile_in,0
        jz @F
        invoke CloseHandle,hfile_in     
@@:     cmp hfile_out,0
        jz @F
        invoke CloseHandle,hfile_out     
@@:     
       ;remove created file if skipped
        cmp filecreated,1
        jnz @F
        cmp _exitvalue,0
        jnz @F
        invoke DeleteFile,heap_cesta2     
@@:
        mov eax,_exitvalue
        ret


_Decrypt_invalidversion:
        inc files_skipped
        push lpFilename
        invoke Pis,file_mem_out,lpstr("ERROR - Unable to determine version of input file: $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_end

_Decrypt_open_err:
        inc files_skipped
        push lpFilename
        invoke GetLastError
        push eax
        invoke Pis,file_mem_out,lpstr("ERROR - Unable to open input file (Error: $h): $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_end

_Decrypt_open2_err:
        inc files_skipped
        push heap_cesta2
        invoke GetLastError
        push eax
        invoke Pis,file_mem_out,lpstr("ERROR - Unable to create output file (Error: $h): $s",NEWLINE)
        invoke AddToLogFile,file_mem_out               
        jmp _Decrypt_end  

_Decrypt_read_err:
        inc files_skipped
        push lpFilename
        invoke GetLastError
        push eax
        invoke Pis,file_mem_out,lpstr("ERROR - Unable to read input file (Error: $h): $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_end

_Decrypt_write_err:
        inc files_skipped
        push heap_cesta2
        invoke GetLastError
        push eax
        invoke Pis,file_mem_out,lpstr("SKIPPED - Unable to write to file (Error: $h): $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_end

_Decrypt_wrong_key:
        inc files_skipped
        push lpFilename
        invoke Pis,file_mem_out,lpstr("SKIPPED - Header doesn't match with loaded keys. (Encrypted with different key): $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_end

_Decrypt_aes_err:
        inc files_skipped
        push lpFilename
        invoke Pis,file_mem_out,lpstr("SKIPPED - Unable to decrypt file (AES internal error): $s",NEWLINE)
        invoke AddToLogFile,file_mem_out
        jmp _Decrypt_end


_DecryptFile endp

reverse_asm_go/simple_reverse_os_stat/simple_reverse_os_stat.go
package main

import (
	"fmt"
	"os"
)

func main() {

	//os* os_Stat(const void* path, int path_len)
	// arg0: rax
	//.text:000000000048A058 lea     rax, "/home/vince"
	//.text:000000000048A05F mov     ebx, 0Bh
	//.text:000000000048A064 call    os_Stat 			-- os* pFileInfo = os_Stat(rax, ebx)

	path, _ := os.UserHomeDir()
	//path := "/home/vince"
	fileInfo, _ := os.Stat(path)

	//.text:000000000048A069 mov     rcx, [rax+18h]		-- rcx = &pFileInfo.IsDir
	//.text:000000000048A06D mov     rax, rbx			-- rax = ???
	//.text:000000000048A070 call    rcx				-- rax = *pFileInfo.IsDir()
	isdir := fileInfo.IsDir()

	// name := fileInfo.Name()
	// fmt.Printf("name = %s", name)
	// size := fileInfo.Size()
	// fmt.Printf("size = %d", size)

	//.text:000000000048A072 test    al, al				-- if rax == 0 goto NOT_A_DIR
	if isdir {
		fmt.Println("path is a dir")
	} else {
		//NOT_A_DIR:
		fmt.Println("path is not a dir")
	}

	fmt.Println("exiting")
}


decrypt.go
////////////////////////////////////////////////////////////////////////////////
// Copyright 2022 Vince Ricosti
// This code is a free and virus free implementation of the ech0raix decoder
// if you have been infected by ech0raix (aka qnap encrypt) and you have paid
// you should have received a zip file containing some binaries for different
// platforms.
// Personnaly even if I have checked that the binary was virus free I prefer to
// execute a well-known code and I advise you to do the same.
// Personal message to the guy who has encrypted my data:
// Karma will come back very hard on you and I wish you will die from a
// rectum's cancer
////////////////////////////////////////////////////////////////////////////////

package main

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strings"
)

//const key = "4STDs9cmUlkiujXuLkdTouoqOIfER4TE"

const _10MB_ = (10 * 1024 * 1024)   //10485760
const _100MB_ = (100 * 1024 * 1024) //104857600

const FILEBUFFERSIZE = _10MB_

var period = -1

func UNUSED(x ...interface{}) {}

func check(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func If[T any](cond bool, vtrue, vfalse T) T {
	if cond {
		return vtrue
	}
	return vfalse
}

func trimFileExtension(fileName string) (filename string, ext string) {
	ext = filepath.Ext(fileName)
	return strings.TrimSuffix(fileName, ext), ext
}

func findSubBufferPos(bigbuf []byte, buf []byte) int64 {
	for i := 0; i <= len(bigbuf)-len(buf); i++ {
		suba := bigbuf[i : i+len(buf)]
		if bytes.Equal(suba, buf) {
			return int64(i)
		}
	}
	return -1
}

func seekRead(inFile *os.File, offset int64, buf []byte) (n int, err error) {
	n = 0
	_, err = inFile.Seek(offset, 0)
	if err == nil {
		n, err = io.ReadFull(inFile, buf)
	}
	return n, err
}

func seekReadCmp(inFile *os.File, offset int64, refBuf []byte) (found bool, err error) {
	buf := make([]byte, len(refBuf))
	bytesread, _ := seekRead(inFile, offset, buf)
	if bytesread == len(refBuf) && bytes.Equal(buf, refBuf) {
		found = true
	}
	return found, err
}

func findPeriod(inFile *os.File, filesize int64, iv []byte) int {
	periodFound := int64(0)

	// Check if period is filesize/2-16
	found, _ := seekReadCmp(inFile, filesize/2, iv)
	periodFound = If(found, filesize/2-16, 0)
	if periodFound <= 0 {
		// Check if period is _10MB_
		found, _ := seekReadCmp(inFile, 16+_10MB_, iv)
		periodFound = If(found, int64(_10MB_), 0)
		if periodFound <= 0 {
			// Check if period is _100MB_
			found, _ := seekReadCmp(inFile, 16+_100MB_, iv)
			periodFound = If(found, int64(_100MB_), 0)
		}
	}

	//We could not find a known period so read 200MB max and try to find one
	if periodFound <= 0 && filesize <= 2*_100MB_+16 {
		tmpbuf := make([]byte, 2*_100MB_+16)
		bytesread, _ := seekRead(inFile, 16, tmpbuf)
		if bytesread > 0 {
			periodFound = findSubBufferPos(tmpbuf, iv)
		}
	}

	return int(periodFound)
}

func findEch0raixMarker(buf []byte) (pos int) {
	//00 00 00 00 00 00 00 65 43 68 30 72 61 69 78
	marker := []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 'e', 'C', 'h', '0', 'r', 'a', 'i', 'x'}
	return int(findSubBufferPos(buf, marker))
}

func decryptFile(inputName string, outputName string, key []byte, deleteCiphered bool) {

	fi, _ := os.Stat(inputName)
	fmt.Printf("Decrypting: %s (%d)\n", inputName, fi.Size())
	if fi.Size() <= 16 {
		fmt.Printf("File is too small %d => skip\n", fi.Size())
		return
	}
	if _, err := os.Stat(outputName); err == nil {
		if fi_out, _ := os.Stat(outputName); fi_out.Size() > 0 {
			fmt.Println("A filename withtout the .encrypt already exists => skip")
			return
		}
	}

	// Open the input and output files
	inFile, err := os.OpenFile(inputName, os.O_RDONLY, 0)
	check(err)
	defer func(inFile *os.File) { _ = inFile.Close() }(inFile)
	outFile, err := os.OpenFile(outputName, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0777)
	check(err)
	defer func(outFile *os.File) { _ = outFile.Close() }(outFile)

	// read first 16 bytes that normally corresponds to iv
	iv := make([]byte, 16)
	_, err = io.ReadFull(inFile, iv[:])
	check(err)

	// find period
	period = findPeriod(inFile, fi.Size(), iv)
	if period < 0 {
		fmt.Printf("Cannot find period of this file => skip\n")
		outFile.Close()
		os.Remove(outputName)
		return
	}

	//bitwiseNotMinus15 := int64(math.MaxInt64 - 15)
	//periodAligned := (period & bitwiseNotMinus15) + 16
	//fmt.Printf("period (16 bytes aligned): %d (original period %d)\n", periodAligned, period)

	// Go back to begin of file
	_, err = inFile.Seek(0, 0)
	check(err)

	// Create the cipher object and decrypt the data
	block, err := aes.NewCipher([]byte(key))
	check(err)

	var stream cipher.Stream = nil
	input_buffer := make([]byte, period)
	decrypted_bytes := make([]byte, period)
	read_len, total_len := 0, int64(0)
	new_period := true
	for ok := true; ok; ok = (read_len > 0) {
		if new_period {
			new_period = false
			read_len, err = io.ReadFull(inFile, iv[:])
			if read_len > 0 && err == nil {
				stream = cipher.NewCFBDecrypter(block, iv)
				total_len += int64(read_len)
			} else {
				fmt.Printf("Could not read iv buffer\n")
			}
		}

		read_len, err = inFile.Read(input_buffer)
		if read_len > 0 && err == nil {
			stream.XORKeyStream(decrypted_bytes, input_buffer)
			markerPos := findEch0raixMarker(decrypted_bytes)
			// ech0raix marker was not found => keep on decrypting
			if markerPos == -1 {
				total_len += int64(read_len)
				_, err = outFile.Write(decrypted_bytes)
				if err == nil {
					total_len, err = inFile.Seek(int64(16+period), 1)
					if err == nil {
						new_period = true
					}
				}
			} else {
				// We have found the ech0raix marker
				total_len += int64(markerPos + 1)
				_, err = outFile.Write(decrypted_bytes[0:markerPos])
				if err == nil {
					inFile.Close()
					outFile.Close()
					fmt.Printf("File was successflully decrypted (period = %d)\n", period)
					if deleteCiphered {
						err = os.Remove(inputName)
						fmt.Printf("Encrypted file removal: %s\n", If(err == nil, "OK", "KO"))
					}
					break
				}
			}
		} else {
			fmt.Printf("Could not read file\n")
		}
	}
}

func main() {

	var key string
	var delete_ciphered_files bool
	var start_dir string

	flag.StringVar(&key, "k", "YOUR_32_BYTES_KEY", "your key")
	flag.BoolVar(&delete_ciphered_files, "d", false, "delete ciphered files")
	flag.StringVar(&start_dir, "s", ".", "directory from where we start to decrypt files")

	flag.Parse()

	if flag.NFlag() < 2 {
		fmt.Println("Usage: ech0raix_decryptor -k YOUR_KEY [-c] -s rootdir")
		fmt.Println("   Example: ech0raix_decryptor -k 4STDs9cmUlkiujXuLkdTouoqOIfER4TE -s .")
		fmt.Println("   Example: ech0raix_decryptor -k 4STDs9cmUlkiujXuLkdTouoqOIfER4TE -s /share/data")
		fmt.Println("   Example: ech0raix_decryptor -d -k 4STDs9cmUlkiujXuLkdTouoqOIfER4TE -s /share/data")
		os.Exit(1)
	}

	if len(key) != 32 {
		fmt.Println("Error: the key should be 32 bytes long")
	}
	if len(start_dir) == 0 {
		fmt.Println("Error: invalid empty path")
	}

	if start_dir[0:1] != "/" {
		tmp_path, _ := filepath.Abs(start_dir)
		flag.Set("s", tmp_path)
	}

	if stat, err := os.Stat(start_dir); err != nil || !stat.IsDir() {
		fmt.Println("Start directory does not exists")
		os.Exit(1)
	}

	filepath.Walk(start_dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Fatalf(err.Error())
		}
		if !info.IsDir() {

			_, filename := filepath.Split(path)
			file_ext := strings.ToLower(filepath.Ext(filename))
			readme_file := strings.ToLower("README_FOR_DECRYPT.txt")
			if strings.HasPrefix(strings.ToLower(filename), readme_file) {
				fmt.Printf("Removing %s\n", path)
				os.Remove(path)
			} else if file_ext == ".encrypt" {
				dstpath := strings.TrimSuffix(path, file_ext)
				decryptFile(path, dstpath, []byte(key), delete_ciphered_files)
			}
		}
		return nil
	})
}
